### PR TITLE
Add metrics to the allocators repair loops

### PR DESCRIPTION
**Pull request by aojea**
/kind documentation
/kind feature

Following up on the last bug we had here https://github.com/kubernetes/kubernetes/pull/120623#issuecomment-1718841396

The repair loops are great for saving us of leaks, but the side effect
is that bugs can go unnoticed for a long time, so we need some
signal to be able to identify those errors proactivily.

Add fours new metrics to identify:
- errors on the reconcile loop on the node port repair loop
- errors on the reconcile loop on the clusterip repair loop
- errors per clusterip
- errors per nodeport
    

```release-note
The kube-apiserver exposes four new metrics to inform about errors on the clusterIP and nodePort allocation logic
```

/sig network

### CODE REVIEW

**Comment by aojea** [comment]
/assign @thockin 

tagging sig-instrumentation for visibility
/cc @logicalhan @dgrisonnet

**Comment by aojea** [comment]
/cc @wojtek-t 

I think he will like this 😄

File: pkg/registry/core/service/ipallocator/controller/metrics.go
Comment (logicalhan) [inline pkg/registry/core/service/ipallocator/controller/metrics.go:27]: generally we've actually just used `apiserver` as the prefix.

Comment (aojea) [reply pkg/registry/core/service/ipallocator/controller/metrics.go]: changed

**Comment by wojtek-t** [comment]
> I think he will like this 😄

Didn't look into the code exporting them, but I'm glad to see the metric :)

**Review by dgrisonnet — APPROVED** [review]
The new metrics look good to me

**Comment by aojea** [comment]
this is really a sig network feature , so Dan is a prefect reviewer for this and we offload Tim 
/assign @danwinship

File: pkg/registry/core/service/ipallocator/controller/repair.go
Comment (danwinship) [inline pkg/registry/core/service/ipallocator/controller/repair.go:143]: same comment on RetryOnConflict

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve 
(from sig inst)

**Comment by logicalhan** [comment]
/hold 

(other comments)

File: pkg/registry/core/service/portallocator/controller/repair.go
Comment (aojea) [reply pkg/registry/core/service/portallocator/controller/repair.go:100]: we want to know if there are conflicts too

File: pkg/registry/core/service/portallocator/controller/metrics.go
Comment (aojea) [reply pkg/registry/core/service/portallocator/controller/metrics.go:55]: > reconcile_errors_total is more or less just the sum of all the port_errors_total values, right?

not really, the only allocator errors that are returned as errors are not going to happen, one is range is full and other is the unknown error, most of the errors returned on the reconcile loop are api operations

File: pkg/registry/core/service/portallocator/controller/repair.go
Comment (aojea) [reply pkg/registry/core/service/portallocator/controller/repair.go:100]: at this stage I'd like to get data of all possible failures and avoid masking some of the errors.In the long term we can correlate the metrics with the behaviors and decide if is worth it or not to increment on conflict errors, it is clear that we had a lot of bugs hidden by the allocators, so I prefer to start with a more granular approach

File: pkg/registry/core/service/ipallocator/controller/metrics.go
Comment (aojea) [reply pkg/registry/core/service/ipallocator/controller/metrics.go:45]: answered above

File: pkg/registry/core/service/ipallocator/controller/repair.go
Comment (aojea) [reply pkg/registry/core/service/ipallocator/controller/repair.go:143]: answerved there

File: pkg/registry/core/service/portallocator/controller/metrics.go
Comment (danwinship) [inline pkg/registry/core/service/portallocator/controller/metrics.go:34]:
```suggestion
	// nodePortRepairPortErrors indicates the number of errors found by the repair loop
	// divided by the type of error:
	// leak, repair, full, outOfRange, duplicate, unknown
```

File: pkg/registry/core/service/portallocator/controller/metrics.go
Comment (danwinship) [inline pkg/registry/core/service/portallocator/controller/metrics.go:40]:
```suggestion
			Help:           "Number of errors detected on ports by the repair loop broken down by type of error: leak, repair, full, outOfRange, duplicate, unknown",
```

File: pkg/registry/core/service/portallocator/controller/metrics.go
Comment (danwinship) [inline pkg/registry/core/service/portallocator/controller/metrics.go:45]:
```suggestion
	// nodePortRepairReconcileErrors indicates the number of times the repair loop has failed to repair
	// the errors it detected.
```

File: pkg/registry/core/service/portallocator/controller/metrics.go
Comment (danwinship) [inline pkg/registry/core/service/portallocator/controller/metrics.go:51]:
```suggestion
			Help:           "Number of reconciliation failures on the nodeport repair reconcile loop",
```

**Review by danwinship — COMMENTED** [review]
suggested rewordings; likewise for the other two metrics

(But if sig instrumentation is OK with these metrics, I have no objections.)

**Comment by aojea** [comment]
> suggested rewordings; likewise for the other two metrics
> 
> (But if sig instrumentation is OK with these metrics, I have no objections.)

/hold cancel

Applied suggestions @danwinship , we need lgtm

File: pkg/registry/core/service/ipallocator/controller/metrics.go
Comment (danwinship) [inline pkg/registry/core/service/ipallocator/controller/metrics.go:40]: sorry, was going to lgtm this but that should say "clusterips" not "ports", right?

File: pkg/registry/core/service/ipallocator/controller/metrics.go
Comment (danwinship) [inline pkg/registry/core/service/ipallocator/controller/metrics.go:33]:
```suggestion
	// divided by the type of error:
```

**Review by danwinship — COMMENTED** [review]
/lgtm
/hold
if you want to fix

File: pkg/registry/core/service/ipallocator/controller/metrics.go
Comment (aojea) [reply pkg/registry/core/service/ipallocator/controller/metrics.go:40]: yep. let me fix it

File: pkg/registry/core/service/ipallocator/controller/metrics.go
Comment (aojea) [reply pkg/registry/core/service/ipallocator/controller/metrics.go]: fixed

**Comment by aojea** [comment]
/hold cancel

ping @danwinship , thanks for finding those typos

**Comment by aojea** [comment]
> E2eNode Suite: [It] [sig-node] OOMKiller [LinuxOnly] [NodeConformance] single process container The containers terminated by OOM killer should have the reason set to OOMKilled expand_less	

https://github.com/kubernetes/kubernetes/issues/119600

/test pull-kubernetes-node-e2e-containerd

**Pull request merged**
